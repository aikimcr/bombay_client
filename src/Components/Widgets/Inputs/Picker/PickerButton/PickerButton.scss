@use '@styles/mixins.scss';

.picker-button {
  display: flex;
  flex-direction: row;
  gap: .5rem;
  justify-content: start;
  align-items: center;
  anchor-name: --picker-button;

  @include mixins.buttonBase($borderRadius: 1rem, $borderStyle: solid, $borderWidth: 1px);

  &:hover:not(:disabled) {
    box-shadow: var(--color-white-200) 2px 2px 1px 1px
  }

  &:active:not(:disabled) {
    box-shadow: var(--color-white-200) 2px 2px 1px 1px inset
  }
}