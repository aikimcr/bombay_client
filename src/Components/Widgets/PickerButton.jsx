import PropTypes from 'prop-types';
import { useRef, useState } from 'react';
import { PickerList } from '.';
import { Button } from './Buttons';

import './OldLabeledInput.scss';
function PickerButton(props) {
  const topRef = useRef(null);
  const inputRef = useRef(null);

  const modelId = props.model ? props.model.id : 'isNew';
  const buttonId = `${props.modelName}-name-${modelId}`;

  const [showPickerList, setShowPickerList] = useState(false);
  const [currentModel, setCurrentModel] = useState(
    props.model ? props.model : null,
  );

  const buttonLabel = currentModel ? currentModel.name : '<Please Choose>';

  const initialValue = currentModel ? currentModel.id : 0;

  function modelPicked(newModel) {
    setCurrentModel(newModel);
    inputRef.current.value = newModel[props.fieldName];
    props.onModelPicked(newModel);
    setShowPickerList(false);
  }

  function pickAModel(evt) {
    evt.preventDefault();
    setShowPickerList((oldShow) => !oldShow);
  }

  return (
    <div
      ref={topRef}
      className="picker-button"
      data-modelname={props.modelName}
      data-targetfield={props.targetField}
    >
      <label htmlFor={buttonId}>{props.labelText}</label>
      <input
        type="number"
        ref={inputRef}
        name={props.targetField}
        defaultValue={initialValue}
      ></input>

      <Button
        id={buttonId}
        text={buttonLabel}
        disabled={false}
        onClick={pickAModel}
      />

      <PickerList
        pickModel={modelPicked}
        isOpen={showPickerList}
        collectionClass={props.collectionClass}
      />
    </div>
  );
}

PickerButton.propTypes = {
  collectionClass: PropTypes.func.isRequired, // To get the list of models
  model: PropTypes.object, // A currently set model (optional)
  modelName: PropTypes.string.isRequired, // Should match collectionClass, mostly a convenience
  fieldName: PropTypes.string.isRequired, // The field from the model used to set the target value
  targetField: PropTypes.string.isRequired, // The key in the data from the form
  labelText: PropTypes.string.isRequired,
  onModelPicked: PropTypes.func.isRequired, // Called with the selected model from the collection
};

export default PickerButton;
