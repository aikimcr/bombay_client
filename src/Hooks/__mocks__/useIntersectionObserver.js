import { useEffect, useRef } from "react";

export class mockEntry {
  constructor(target, isIntersecting = false) {
    this.target = target;
    this.isIntersecting = isIntersecting;
    this.isVisible = false;
  }
}

export class mockObserver {
  static observers = [];

  constructor(callback, options) {
    this.callback = callback;
    this.options = options;
    this.entries = [];
    this.constructor.observers.push(this);
  }

  observe(target) {}

  unobserve(target) {}

  disconnect() {}

  _fireIntersect(target, isIntersecting = true) {
    const entry = new mockEntry(target, isIntersecting);
    this.callback([entry], this);
  }
}

export function useIntersectionObserver(topRef, callback, options = {}) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) return;

    if (observer.current == null) {
      observer.current = new mockObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [observer.current, topRef]);

  return observer;
}

export default useIntersectionObserver;
