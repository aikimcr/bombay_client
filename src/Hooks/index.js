export * from './useLoginTracking';
export * from './useModelCollection';
