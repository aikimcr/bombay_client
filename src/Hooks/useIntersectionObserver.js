import { useEffect, useRef } from "react";

export function useIntersectionObserver(topRef, callback, options = {}) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) return;

    if (observer.current == null) {
      observer.current = new IntersectionObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [callback, options, topRef]);

  return observer;
}

export default useIntersectionObserver;
