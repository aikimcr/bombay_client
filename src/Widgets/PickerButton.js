import PropTypes from 'prop-types';
import { useRef, useState } from 'react';
import PickerList from './PickerList';
import Button from './Button';

import './LabeledInput.scss';

function PickerButton(props) {
    const topRef = useRef(null);

    const modelId = props.model ? props.model.get('id') : 'isNew';
    const buttonId = `${props.modelName}-name-${modelId}`;

    const [showPickerList, setShowPickerList] = useState(false);
    const [currentModel, setCurrentModel] = useState(props.model ? props.model : null);
    
    const buttonLabel = currentModel ?
                        currentModel.get('name') :
                        '<Please Choose>';

    function modelPicked(newModel) {
        setCurrentModel(newModel);
        props.onModelPicked(newModel);
        setShowPickerList(false);
    }

    function pickAModel(evt) {
        evt.preventDefault();
        setShowPickerList(true);
    }

    return (
        <div 
            ref={topRef}
            className='picker-button'
            data-modelname={props.modelName}
        >
            <label htmlFor={buttonId}>{props.labelText}</label>
            <Button
                id={buttonId}
                label={buttonLabel}
                disabled={false}
                onClick={pickAModel}
            />
            <PickerList
                pickModel={modelPicked}
                isOpen={showPickerList}
                collectionClass={props.collectionClass}
            />
        </div>
    );
}

PickerButton.propTypes = {
    modelName: PropTypes.string.isRequired,
    labelText: PropTypes.string.isRequired,
    collectionClass: PropTypes.func.isRequired,
    onModelPicked: PropTypes.func.isRequired,
    model: PropTypes.object,
}

export default PickerButton;