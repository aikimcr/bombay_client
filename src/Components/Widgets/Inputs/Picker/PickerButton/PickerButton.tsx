import React from 'react';

import './PickerButton.scss';

interface PickerButtonProps {
  id?: string;
  buttonText: string;
  disabled?: boolean;
  openState: boolean;
  openStateToggle: () => void;
  popoverTarget: string;
}

export const PickerButton: React.FC<PickerButtonProps> = ({
  id = 'picker-button',
  buttonText,
  disabled = false,
  openState,
  openStateToggle,
  popoverTarget,
}): React.ReactElement => {
  return (
    <button
      id={id}
      className="picker-button"
      data-testid="picker-button"
      type="button"
      onClick={openStateToggle}
      aria-haspopup="menu"
      aria-expanded={openState}
      aria-controls={popoverTarget}
      popoverTarget={popoverTarget}
      popoverTargetAction="toggle"
      disabled={disabled}
    >
      {buttonText}
      {openState ? (
        <i className="fa-solid fa-caret-up"></i>
      ) : (
        <i className="fa-solid fa-caret-down"></i>
      )}
    </button>
  );
};
