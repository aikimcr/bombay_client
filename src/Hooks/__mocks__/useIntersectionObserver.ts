import { useEffect, useRef, RefObject } from 'react';

import { IntersectionObserverOptions } from '../useIntersectionObserver';
export class mockEntry {
  boundingClientRect: DOMRectReadOnly;
  intersectionRatio: number;
  intersectionRect: DOMRectReadOnly;
  isIntersecting: boolean;
  rootBounds: DOMRectReadOnly;
  target: Element;
  time: number;
  isVisible: boolean;

  constructor(target: Element, isIntersecting = false) {
    this.target = target;
    this.isIntersecting = isIntersecting;
    this.isVisible = false;
  }
}

export class mockObserver {
  root: Element;
  rootMargin: string;
  thresholds: readonly number[];
  callback: IntersectionObserverCallback;
  options: IntersectionObserverOptions;
  entries: mockEntry[];

  static observers: mockObserver[] = [];

  constructor(
    callback: IntersectionObserverCallback,
    options: IntersectionObserverOptions,
  ) {
    this.callback = callback;
    this.options = options;
    this.entries = [];
    mockObserver.observers.push(this);
  }

  observe(_target: Element) {}

  unobserve(_target: Element) {}

  takeRecords() {
    return this.entries;
  }

  disconnect() {}

  _fireIntersect(target: Element, isIntersecting = true) {
    const entry = new mockEntry(target, isIntersecting);
    this.callback([entry], this);
  }
}

export function useIntersectionObserver(
  topRef: RefObject<Element>,
  callback: IntersectionObserverCallback,
  options: IntersectionObserverOptions = {},
) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) {
      return;
    }

    if (observer.current == null) {
      observer.current = new mockObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [observer.current, topRef]);

  return observer;
}

export default useIntersectionObserver;
