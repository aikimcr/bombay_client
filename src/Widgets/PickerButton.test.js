import { act, render } from '@testing-library/react';

import * as Network from '../Network/Network';
import * as mockObserver from '../Hooks/useIntersectionObserver';

import * as Login from '../Network/Login';
jest.mock('../Network/Login');

import ModelBase from '../model/ModelBase';
import CollectionBase from '../model/CollectionBase';

class TestModel extends ModelBase { };

class TestCollection extends CollectionBase {
    constructor(options = {}) {
        super('/table1', { ...options, modelClass: TestModel });
    }
}

import PickerButton from './PickerButton';

jest.useFakeTimers();

const testToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJGREM4MTEzOCIsInVzZXIiOnsiaWQiOjEsIm5hbWUiOiJhZG1pbiIsImFkbWluIjpmYWxzZX0sImlhdCI6MTY2NTk2NTA5OX0.2vz14X7Tm-oFlyOa7dcAF-5y5ympi_UlWyJNxO4xyS4';

beforeEach(() => {
    localStorage.setItem('jwttoken', testToken);
})

afterEach(() => {
    while (mockObserver.mockObserver.observers.length > 0) {
        mockObserver.mockObserver.observers.pop();
    }

    localStorage.removeItem('jwttoken');
})

it('should show the button', async () => {
    const loginPromise = Login._setupMocks();
    loginPromise.resolve({
        loggedIn: true,
        token: testToken,
    });

    const onModelPicked = jest.fn();

    const { asFragment } = render(
        <PickerButton
            collectionClass={TestCollection}
            modelName='table1'
            fieldName='id'
            targetField='table2_id'
            labelText='Pick A Model'
            onModelPicked={onModelPicked}
        />
    );

    expect(asFragment).toMatchSnapshot();

    expect(mockObserver.mockObserver.observers.length).toBe(0);
})

it('should show the list on click', async () => {
    const loginPromise = Login._setupMocks();
    loginPromise.resolve({
        loggedIn: true,
        token: testToken,
    });

    const onModelPicked = jest.fn();

    const { resolve } = Network._setupMocks();

    const { container, queryByText } = render(
        <PickerButton
            collectionClass={TestCollection}
            modelName='table1'
            fieldName='id'
            targetField='table2_id'
            labelText='Pick A Model'
            onModelPicked={onModelPicked}
        />
    );

    let listComponent = container.querySelector('.picker-component');
    expect(listComponent).toBeNull;
    expect(mockObserver.mockObserver.observers.length).toBe(0);

    const showButton = queryByText(/Please Choose/);
    const [fetchBody] = makeModels(10, {});

    await act(async () => {
        showButton.click();
        resolve(fetchBody);
    })

    expect(mockObserver.mockObserver.observers.length).toBe(1);
    listComponent = container.querySelector('.picker-component');
    expect(listComponent).toBeInTheDocument();
})

it('should call the callback and close the list when an item is clicked', async () => {
    const loginPromise = Login._setupMocks();
    loginPromise.resolve({
        loggedIn: true,
        token: testToken,
    });

    const onModelPicked = jest.fn();

    const { resolve } = Network._setupMocks();

    const { container, queryByText } = render(
        <PickerButton
            collectionClass={TestCollection}
            modelName='table1'
            fieldName='id'
            targetField='table2_id'
            labelText='Pick A Model'
            onModelPicked={onModelPicked}
        />
    );

    const showButton = queryByText(/Please Choose/);
    const [fetchBody, models] = makeModels(10, {}, undefined, (def) => {
        def.table2_id = parseInt(Math.random() * 100);
    });

    await act(async () => {
        showButton.click();
        resolve(fetchBody);
    })

    expect(onModelPicked).not.toBeCalled();

    expect(mockObserver.mockObserver.observers.length).toBe(1);
    const el = queryByText(models[2].get('name'));

    await act(async () => {
        el.click();
    });

    expect(onModelPicked).toBeCalled();
    expect(onModelPicked).toBeCalledWith(models[2]);
    const listComponent = container.querySelector('.picker-component');
    expect(listComponent).toBeNull;
})

it('should close the list without calling if the button is pushed again', async () => {
    const loginPromise = Login._setupMocks();
    loginPromise.resolve({
        loggedIn: true,
        token: testToken,
    });

    const onModelPicked = jest.fn();

    const { resolve } = Network._setupMocks();

    const { container, queryByText } = render(
        <PickerButton
            collectionClass={TestCollection}
            modelName='table1'
            targetField='table2_id'
            fieldName='table2_id'
            labelText='Pick A Model'
            onModelPicked={onModelPicked}
        />
    );

    const showButton = queryByText(/Please Choose/);
    const [fetchBody, models] = makeModels(10, {});

    await act(async () => {
        showButton.click();
        resolve(fetchBody);
    })

    expect(onModelPicked).not.toBeCalled();

    expect(mockObserver.mockObserver.observers.length).toBe(1);
    let listComponent = container.querySelector('.picker-component');
    expect(listComponent).toBeInTheDocument();

    expect(onModelPicked).not.toBeCalled();

    await act(async () => {
        showButton.click();
    })

    listComponent = container.querySelector('.picker-component');
    expect(listComponent).not.toBeInTheDocument();
});