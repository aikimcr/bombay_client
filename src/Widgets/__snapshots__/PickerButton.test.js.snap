// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show the button 1`] = `[Function]`;
